File: samples/nodejs/museum-exhibit-studio/src/service.ts
import {
  CopilotClient,
  type PermissionHandler,
  type SessionConfig,
} from "@github/copilot-sdk";
import { buildExhibitPrompt, systemMessage } from "./prompts.js";
import {
  buildResearchPrompt,
  incompleteResearch,
  maximumResearchArticleReads,
  maximumResearchSearchCalls,
  parseResearchResult,
  researchSystemMessage,
  researchTimeoutMs,
  type ResearchResult,
} from "./research.js";
import { validateExhibit, type ExhibitValidation } from "./validator.js";

export const generationTimeoutMs = 120_000;

export interface CuratorSession {
  sendAndWait(prompt: string, timeout?: number): Promise<
    { data: { content: string } } | undefined
  >;
  disconnect(): Promise<void>;
}

export interface CuratorClient {
  start(): Promise<void>;
  createSession(configuration: SessionConfig): Promise<CuratorSession>;
  stop(): Promise<unknown>;
}

export interface GeneratedExhibit {
  content: string;
  validation: ExhibitValidation;
}

export function createSessionConfiguration(model?: string): SessionConfig {
  return {
    clientName: "museum-exhibit-studio",
    model: model?.trim() || undefined,
    availableTools: [],
    streaming: false,
    systemMessage: {
      mode: "replace",
      content: systemMessage,
    },
  };
}

export function createWikipediaPermissionHandler(): PermissionHandler {
  let searchCalls = 0;
  let articleReads = 0;

  return (request) => {
    if (request.kind === "mcp" && request.serverName === "wikipedia") {
      if (["search", "wikipedia-search"].includes(request.toolName) &&
          articleReads === 0 && searchCalls < maximumResearchSearchCalls) {
        searchCalls += 1;
        return { kind: "approve-once" };
      }
      if (["readArticle", "wikipedia-readArticle"].includes(request.toolName) &&
          searchCalls > 0 && articleReads < maximumResearchArticleReads) {
        articleReads += 1;
        return { kind: "approve-once" };
      }
    }
    return {
      kind: "reject",
      feedback:
        "This session permits at most 5 Wikipedia searches followed by one article retrieval.",
    };
  };
}

export function createResearchSessionConfiguration(model?: string): SessionConfig {
  return {
    clientName: "museum-exhibit-studio-research",
    model: model?.trim() || undefined,
    streaming: false,
    systemMessage: {
      mode: "replace",
      content: researchSystemMessage,
    },
    availableTools: ["wikipedia-search", "wikipedia-readArticle"],
    mcpServers: {
      wikipedia: {
        command: "npx",
        args: ["-y", "wikipedia-mcp@1.0.3"],
        workingDirectory: process.cwd(),
        tools: ["search", "readArticle"],
      },
    },
    onPermissionRequest: createWikipediaPermissionHandler(),
  };
}

export class MuseumExhibitService {
  constructor(private readonly client: CuratorClient) {}

  async research(approvedFacts: Iterable<string>, model?: string): Promise<ResearchResult> {
    const { facts, prompt } = buildResearchPrompt(approvedFacts);
    let session: CuratorSession | undefined;

    try {
      await this.client.start();
      session = await this.client.createSession(createResearchSessionConfiguration(model));
      const response = await session.sendAndWait(prompt, researchTimeoutMs);
      const content = response?.data.content;
      if (!content?.trim()) throw new Error("The researcher returned no result.");
      return parseResearchResult(content, facts);
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      return incompleteResearch(facts, message);
    } finally {
      try {
        await session?.disconnect();
      } finally {
        await this.client.stop();
      }
    }
  }

  async generate(approvedFacts: Iterable<string>, model?: string): Promise<GeneratedExhibit> {
    const prompt = buildExhibitPrompt(approvedFacts);
    let session: CuratorSession | undefined;

    try {
      await this.client.start();
      session = await this.client.createSession(createSessionConfiguration(model));
      const response = await session.sendAndWait(prompt, generationTimeoutMs);
      const content = response?.data.content;
      if (!content?.trim()) throw new Error("The curator returned no exhibit content.");
      return { content, validation: validateExhibit(content) };
    } finally {
      try {
        await session?.disconnect();
      } finally {
        await this.client.stop();
      }
    }
  }
}

export function createCopilotCuratorClient(): CuratorClient {
  return new CopilotClient();
}

File: finished/go/museum-exhibit-studio/service.go
package main

import (
	"context"
	"errors"
	"fmt"
	"strings"
	"time"

	copilot "github.com/github/copilot-sdk/go"
	"github.com/github/copilot-sdk/go/rpc"
)

const generationTimeout = 120 * time.Second

type curatorSession interface {
	SendAndWait(context.Context, string) (string, error)
	Disconnect() error
}

type curatorClient interface {
	Start(context.Context) error
	CreateSession(context.Context, *copilot.SessionConfig) (curatorSession, error)
	Stop() error
}

type generatedExhibit struct {
	Content    string
	Validation ExhibitValidation
}

type museumExhibitService struct {
	client curatorClient
}

func createSessionConfiguration(model string) *copilot.SessionConfig {
	return &copilot.SessionConfig{
		ClientName:     "museum-exhibit-studio",
		Model:          strings.TrimSpace(model),
		AvailableTools: []string{},
		Streaming:      copilot.Bool(false),
		SystemMessage: &copilot.SystemMessageConfig{
			Mode:    "replace",
			Content: curatorSystemMessage,
		},
	}
}

func createResearchSessionConfiguration(model string) *copilot.SessionConfig {
	return &copilot.SessionConfig{
		ClientName: "museum-exhibit-studio-research",
		Model:      strings.TrimSpace(model),
		Streaming:  copilot.Bool(false),
		SystemMessage: &copilot.SystemMessageConfig{
			Mode:    "replace",
			Content: researchSystemMessage,
		},
		AvailableTools:      []string{"wikipedia-search", "wikipedia-readArticle"},
		OnPermissionRequest: newWikipediaPermissionHandler(),
		MCPServers: map[string]copilot.MCPServerConfig{
			"wikipedia": copilot.MCPStdioServerConfig{
				Command:          "npx",
				Args:             []string{"-y", "wikipedia-mcp@1.0.3"},
				WorkingDirectory: ".",
				Tools:            []string{"search", "readArticle"},
			},
		},
	}
}

func newWikipediaPermissionHandler() copilot.PermissionHandlerFunc {
	var searchCalls int
	var articleCalls int
	return func(
		request copilot.PermissionRequest,
		_ copilot.PermissionInvocation,
	) (rpc.PermissionDecision, error) {
		var mcpRequest copilot.PermissionRequestMCP
		switch value := request.(type) {
		case copilot.PermissionRequestMCP:
			mcpRequest = value
		case *copilot.PermissionRequestMCP:
			mcpRequest = *value
		default:
			return rejectWikipediaPermission(), nil
		}
		if mcpRequest.ServerName != "wikipedia" ||
			(mcpRequest.ManagedApprovalRequired != nil && *mcpRequest.ManagedApprovalRequired) {
			return rejectWikipediaPermission(), nil
		}

		switch mcpRequest.ToolName {
		case "search", "wikipedia-search":
			if searchCalls >= 1 {
				return rejectWikipediaPermission(), nil
			}
			searchCalls++
		case "readArticle", "wikipedia-readArticle":
			if searchCalls == 0 || articleCalls >= 1 {
				return rejectWikipediaPermission(), nil
			}
			articleCalls++
		default:
			return rejectWikipediaPermission(), nil
		}
		return &rpc.PermissionDecisionApproveOnce{}, nil
	}
}

func rejectWikipediaPermission() rpc.PermissionDecision {
	feedback := "This workshop permits only read-only Wikipedia search and article retrieval."
	return &rpc.PermissionDecisionReject{Feedback: &feedback}
}

func (service museumExhibitService) Generate(ctx context.Context, approvedFacts []string, model string) (result generatedExhibit, err error) {
	prompt, err := buildExhibitPrompt(approvedFacts)
	if err != nil {
		return result, err
	}

	defer func() { err = errors.Join(err, service.client.Stop()) }()
	if err = service.client.Start(ctx); err != nil {
		return result, err
	}

	session, err := service.client.CreateSession(ctx, createSessionConfiguration(model))
	if err != nil {
		return result, err
	}
	defer func() { err = errors.Join(err, session.Disconnect()) }()

	generationContext, cancel := context.WithTimeout(ctx, generationTimeout)
	defer cancel()
	content, err := session.SendAndWait(generationContext, prompt)
	if err != nil {
		return result, err
	}
	if strings.TrimSpace(content) == "" {
		return result, fmt.Errorf("the curator returned no exhibit content")
	}
	return generatedExhibit{Content: content, Validation: validateExhibit(content)}, nil
}

type copilotCuratorClient struct {
	client *copilot.Client
}

func newCopilotCuratorClient() *copilotCuratorClient {
	return &copilotCuratorClient{client: copilot.NewClient(&copilot.ClientOptions{LogLevel: "error"})}
}

func (client *copilotCuratorClient) Start(ctx context.Context) error {
	return client.client.Start(ctx)
}

func (client *copilotCuratorClient) CreateSession(ctx context.Context, config *copilot.SessionConfig) (curatorSession, error) {
	session, err := client.client.CreateSession(ctx, config)
	if err != nil {
		return nil, err
	}
	return copilotCuratorSession{session: session}, nil
}

func (client *copilotCuratorClient) Stop() error {
	return client.client.Stop()
}

type copilotCuratorSession struct {
	session *copilot.Session
}

func (session copilotCuratorSession) SendAndWait(ctx context.Context, prompt string) (string, error) {
	response, err := session.session.SendAndWait(ctx, copilot.MessageOptions{Prompt: prompt})
	if err != nil || response == nil {
		return "", err
	}
	message, ok := response.Data.(*copilot.AssistantMessageData)
	if !ok {
		return "", nil
	}
	return message.Content, nil
}

func (session copilotCuratorSession) Disconnect() error {
	return session.session.Disconnect()
}
